### ISSUE TITLE
Cannot find brokers in response metadata when using python kafka client in docker compose

### ISSUE THREAD

**Issue by ParikhKadam** [bitnami/bitnami-docker-kafka#190]
**Description**

I have a kubernetes cluster setup where there are mainly three services: kafka, zookeeper and producer. The producer is like a python script that uses kafka-python or confluent-kafka library. I have tried both these libraries and both of them faces the same issue with producing messages, hence I assume the issue is in kafka server.

Now, the issue is, when producing messages, I am getting the below error:
```
org.apache.kafka.common.errors.InvalidReplicationFactorException: Replication factor: 1 larger than available brokers: 0.
[2021-09-06 11:26:28,361] INFO [Admin Manager on Broker 1]: Error processing create topic request CreatableTopic(name='mytopic', numPartitions=1, replicationFactor=1, assignments=[], configs=[]) (kafka.server.ZkAdminManager)
org.apache.kafka.common.errors.InvalidReplicationFactorException: Replication factor: 1 larger than available brokers: 0.
```

I have checked that there are brokers live by using the `ls /brokers/ids` in zookeeper shell. Also, I am only facing this issue when producing messages via the producer service written in python. When I create a new topic via kafka-topics from inside the container, the topic is created successfully.

So, I think the issue here lies within the **advertised listeners** configuration but I think I have configured it poperly.

**Steps to reproduce the issue:**

1. Start all the three services in docker-compose or kubernetes.
2. Sample code of the producer service:
```
from confluent_kafka import Producer

conf = {'bootstrap.servers': 'kafka-1:9092', 'client.id': 'test', 'default.topic.config': {'acks': 'all'}}
producer = Producer(conf)
producer.produce("TOPIC", key='key', value='value')
producer.flush()
```
3. Docker compose for zookeeper and kafka:
```
  zookeeper-1:
    image: bitnami/zookeeper:latest
    environment:
      - ZOO_MY_ID=1
      - ALLOW_ANONYMOUS_LOGIN=true
    ports:
      - "2181:2181"
  kafka-1:
    image: bitnami/kafka:latest
    depends_on:
      - zookeeper-1
    ports:
      - "9092:9092"
    environment:
      - KAFKA_BROKER_ID=1
      - KAFKA_CFG_ZOOKEEPER_CONNECT=zookeeper-1:2181
      - KAFKA_CFG_AUTO_CREATE_TOPICS_ENABLE=true
      - KAFKA_CFG_LISTENERS=PLAINTEXT://kafka-1:9092
      - KAFKA_CFG_ADVERTISED_LISTENERS=PLAINTEXT://kafka-1:9092
      - ALLOW_PLAINTEXT_LISTENER=yes
      - KAFKA_CFG_DEFAULT_REPLICATION_FACTOR=1
```

I've also checked the advertised listeners in `server.properties` file of the running container and it is set to `kafka-1:9092`.

**Describe the results you received:**
1. Zookeeper starts.
2. Kafka starts.
3. Producer starts, but when producing the message it gets the error listed above in `server.log` file inside kafka container. 

**Describe the results you expected:**

1. Zookeeper starts.
2. Kafka starts.
3. Producer starts, and produces the message.

**Version**

- Output of `docker version`:

```
Client: Docker Engine - Community
 Version:           20.10.6
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        370c289
 Built:             Fri Apr  9 22:47:17 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.6
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       8728dd2
  Built:            Fri Apr  9 22:45:28 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.4
  GitCommit:        05f951a3781f4f2c1911b05e61c160e9c30eaa8e
 runc:
  Version:          1.0.0-rc93
  GitCommit:        12644e614e25b05da6fd08a38ffa0cfe1903fdec
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

- Output of `docker info`:

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.5.1-docker)
  scan: Docker Scan (Docker Inc., v0.7.0)

Server:
 Containers: 18
  Running: 1
  Paused: 0
  Stopped: 17
 Images: 467
 Server Version: 20.10.6
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc io.containerd.runc.v2 io.containerd.runtime.v1.linux
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 05f951a3781f4f2c1911b05e61c160e9c30eaa8e
 runc version: 12644e614e25b05da6fd08a38ffa0cfe1903fdec
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
 Kernel Version: 5.11.0-27-generic
 Operating System: Ubuntu 20.04.3 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 15.59GiB
 Name: fxpc
 ID: E2TM:O4X3:SHKS:Z6Y2:QVRK:XGCB:X6DB:YR7C:6XC6:TAVA:TPFO:VD3Z
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

- Output of `docker-compose version` (if applicable):

```
docker-compose version 1.28.2, build 67630359
docker-py version: 4.4.1
CPython version: 3.7.9
OpenSSL version: OpenSSL 1.1.0l  10 Sep 2019
```

**Additional environment details (AWS, VirtualBox, Docker for MAC, physical, etc.):**
Physical system running on Ubuntu 20.04.

**Comment by juan131** [comment]
Hi @ParikhKadam 

How are you deploying your solution on Kubernetes? Are you using the [Bitnami Kafka chart](https://github.com/bitnami/charts/tree/master/bitnami/kafka) that configures both Kafka and Zookeeper to run in K8s properly?

```console
$ helm repo add bitnami https://charts.bitnami.com/bitnami
$ helm install my-release bitnami/kafka
```

The chart installation notes will provide you with examples to access the brokers using a producer client inside the cluster (in a similar way you described).

**Comment by ParikhKadam** [comment]
Hello @juan131 

I ain't using helm. I created the kubernetes files using kompose. But I really think the problem is not with kubernetes. Still, will cross verify and revert.

Here are the k8s files generated:
**kafka-1-deployment.yaml**
```
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    kompose.cmd: kompose -f docker-compose.local.yml convert -o ./kubernetes.local/
    kompose.version: 1.22.0 (955b78124)
  creationTimestamp: null
  labels:
    io.kompose.service: kafka-1
  name: kafka-1
spec:
  replicas: 1
  selector:
    matchLabels:
      io.kompose.service: kafka-1
  strategy: {}
  template:
    metadata:
      annotations:
        kompose.cmd: kompose -f docker-compose.local.yml convert -o ./kubernetes.local/
        kompose.version: 1.22.0 (955b78124)
      creationTimestamp: null
      labels:
        io.kompose.service: kafka-1
    spec:
      containers:
        - env:
            - name: ALLOW_PLAINTEXT_LISTENER
              value: "yes"
            - name: KAFKA_BROKER_ID
              value: "1"
            - name: KAFKA_CFG_ADVERTISED_LISTENERS
              value: PLAINTEXT://kafka-1:9092
            - name: KAFKA_CFG_AUTO_CREATE_TOPICS_ENABLE
              value: "true"
            - name: KAFKA_CFG_DEFAULT_REPLICATION_FACTOR
              value: "1"
            - name: KAFKA_CFG_LISTENERS
              value: PLAINTEXT://:9092
            - name: KAFKA_CFG_ZOOKEEPER_CONNECT
              value: zookeeper-1:2181
          image: bitnami/kafka:latest
          name: kafka-1
          ports:
            - containerPort: 9092
          resources: {}
      restartPolicy: Always
status: {}
```

**kafka-1-service.yaml**
```
apiVersion: v1
kind: Service
metadata:
  annotations:
    kompose.cmd: kompose -f docker-compose.local.yml convert -o ./kubernetes.local/
    kompose.version: 1.22.0 (955b78124)
  creationTimestamp: null
  labels:
    io.kompose.service: kafka-1
  name: kafka-1
spec:
  ports:
    - name: "9092"
      port: 9092
      targetPort: 9092
  selector:
    io.kompose.service: kafka-1
status:
  loadBalancer: {}
```

**zookeeper-1-deployment.yaml**
```
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    kompose.cmd: kompose -f docker-compose.local.yml convert -o ./kubernetes.local/
    kompose.version: 1.22.0 (955b78124)
  creationTimestamp: null
  labels:
    io.kompose.service: zookeeper-1
  name: zookeeper-1
spec:
  replicas: 1
  selector:
    matchLabels:
      io.kompose.service: zookeeper-1
  strategy: {}
  template:
    metadata:
      annotations:
        kompose.cmd: kompose -f docker-compose.local.yml convert -o ./kubernetes.local/
        kompose.version: 1.22.0 (955b78124)
      creationTimestamp: null
      labels:
        io.kompose.service: zookeeper-1
    spec:
      containers:
        - env:
            - name: ALLOW_ANONYMOUS_LOGIN
              value: "true"
            - name: ZOO_MY_ID
              value: "1"
          image: bitnami/zookeeper:latest
          name: zookeeper-1
          ports:
            - containerPort: 2181
          resources: {}
      restartPolicy: Always
status: {}
```

**zookeeper-1-service.yaml**
```
apiVersion: v1
kind: Service
metadata:
  annotations:
    kompose.cmd: kompose -f docker-compose.local.yml convert -o ./kubernetes.local/
    kompose.version: 1.22.0 (955b78124)
  creationTimestamp: null
  labels:
    io.kompose.service: zookeeper-1
  name: zookeeper-1
spec:
  ports:
    - name: "2181"
      port: 2181
      targetPort: 2181
  selector:
    io.kompose.service: zookeeper-1
status:
  loadBalancer: {}
```

**Comment by ParikhKadam** [comment]
Hi @juan131 

Seems like you are right. There is some issue with kubernetes. I've tested this and it works fine with docker-compose, but not with k8s.

Will try using helm now and revert..

Thanks

**Comment by juan131** [comment]
Hi @ParikhKadam 

There's something odd in your k8s manifests:

- First, you shouldn't be using a Deployment but a Statefulset instead since the former doesn't guarantee you stable network identifiers.
- Second, you shouldn't hardcore the env. variable below since you won't be able to scale the solution:

```
            - name: KAFKA_CFG_ADVERTISED_LISTENERS
              value: PLAINTEXT://kafka-1:9092
```

I highly recommend you to use the chart since we already gave response to all these details there. If you're not using Helm to deploy your manifests in your cluster (for whatever reason), you can always use:

```console
helm template kafka bitnami/kafka
```

That will render the manifests of the charts and give you pure YAML manifests that you can later on install in your cluster with you preferred tool.

**Comment by ParikhKadam** [comment]
Hi @juan131 

I looked into helm, actually at the template of kafka in helm. It seems to use Statefulset like you mentioned along with headless services for both zookeeper and kafka. Also, it uses a dynamic value for `KAFKA_CFG_ADVERTISED_LISTENERS`.

I wasn't very aware of these concepts at that time, but now I know why they are used. Basically, in multi node kafka cluster, all nodes should be reachable by all other nodes and all nodes must be uniquely identifiable.

Hence, we use statefulset so that all the nodes can store their metadata (especially the broker id) that identifies them uniquely. In case of failure, a new broker is fired by k8s which is mounted with the volume of the failed kafka broker so that they share them metadata and hence, they are ideally the same in kafka ecosystem.

Now, we also need leader election and such things in kafka. For that, we need each node to be able to reach every other node in the cluster and vice versa. For that, instead of using a simple service in k8s, we use headless service which maps DNS to IP for every single kafka node, instead of assigning a Cluster IP to the service itself.

This is what I have understood so far. But I am doubting a thing. All these is required in case where I have multiple kafka nodes, which will surely be the case in production.

But during development, on a local machine, I'll only have  a single kafka and a single zookeeper node. In that case, why doesn't the above config work well? I am sure that something is wrong with k8s config itself, but I don't get what exactly is the issue.

Should I try replacing `PLAINTEXT://kafka-1:9092` with `PLAINTEXT://$(MY_POD_NAME).kafka-headless.default.svc.cluster.local:9092` and use a headless service? I think I should try this as mostly the service which acts as a proxy and load balance might be causing some networking issues. I'll have to look on how to use headless services with deployments.

I personally do not think that use of deployment instead of statefulset is the reason behind this issue, as according to the understanding that I gained, it is only need to maintain state of kafka brokers when one of them fails.

Sorry if I am sounding naive, but I want to understand what's causing this issue rather than simply using helm template to solve it.

**Comment by juan131** [comment]
Hi @ParikhKadam 

> Basically, in multi node kafka cluster, all nodes should be reachable by all other nodes and all nodes must be uniquely identifiable.
> Hence, we use statefulset so that all the nodes can store their metadata (especially the broker id) that identifies them uniquely. In case of failure, a new broker is fired by k8s which is mounted with the volume of the failed kafka broker so that they share them metadata and hence, they are ideally the same in kafka ecosystem.

100% correct. Also, with a statefulset the hostnames to access each Kafka broker individually are stable. I mean, they're sth like `kafka-0`, `kafka-1`, etc. instead of `kafka-XXX` where XXX is a random string and changes if the pod is rescheduled.

> But during development, on a local machine, I'll only have a single kafka and a single zookeeper node. In that case, why doesn't the above config work well? I am sure that something is wrong with k8s config itself, but I don't get what exactly is the issue.
> Should I try replacing PLAINTEXT://kafka-1:9092 with PLAINTEXT://$(MY_POD_NAME).kafka-headless.default.svc.cluster.local:9092 and use a headless service? I think I should try this as mostly the service which acts as a proxy and load balance might be causing some networking issues. I'll have to look on how to use headless services with deployments.

If you're using a single Kafka node (and a single Zookeeper node) you can use simple services (no need of a headless service) for Kafka & Zookeeper. Why? Simple, the service redirects the traffic (round-robin fashion) between every pod matching certain labels (the ones you set in the `selector` field) but in your case there'll be only one pod so that's not a problem.

Could you try with this simplified version of the environment for Kafka (let's rely on the container's default to manage the listener configuration)?

```yaml
spec:
  ...
  template:
    spec:
      containers:
        - env:
            - name: KAFKA_CFG_ZOOKEEPER_CONNECT
              value: zookeeper-1
            - name: ALLOW_PLAINTEXT_LISTENER
              value: "yes"
```

**Comment by ParikhKadam** [comment]
Hi @juan131 

I didn't get time to check this. Will do in a few days and revert.

Mostly, I don't think it would work. Basically, what I have found is, when using `KAFKA_CFG_ADVERTISED_LISTENERS=PLAINTEXT://:9092` which I think is the default one, is used the broker's k8s node's name as the DNS name of broker. And when I specify `KAFKA_CFG_ADVERTISED_LISTENERS=PLAINTEXT://kafka-1:9092`, it uses `kafka-1`as the DNS name of broker. I found this info from the kafka logs. 

So, in the 2nd case, if broker is configured on `kafka-1` then a normal k8s service should work, but still it doesn't.

Still will dig deeper into this and revert. Thanks..
